### ISSUE TITLE
dulwich pull is prone to file lock failures on windows

### ISSUE THREAD

**Issue by 3ll3d00d** [3ll3d00d/beqdesigner#222]
a simple test script

```
from multiprocessing import Process


def g():
    from dulwich import porcelain
    print('pulling')
    porcelain.pull('c:/Users/mattk/.test', 'https://github.com/3ll3d00d/test.git')


def f():
    from dulwich import porcelain
    print('fetching')
    porcelain.fetch('c:/Users/mattk/.test', 'https://github.com/3ll3d00d/test.git')


if __name__ == '__main__':
    for i in range(10):
        targ = f if i % 2 == 0 else g
        p=Process(target=targ)
        p.start()
        p.join()
        print(f'done {i} {p}')
```

will more often than not fail with

```
Traceback (most recent call last):
  File "C:\Users\mattk\Anaconda3_64\envs\beq\lib\multiprocessing\process.py", line 258, in _bootstrap
    self.run()
  File "C:\Users\mattk\Anaconda3_64\envs\beq\lib\multiprocessing\process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\mattk\github\beq\src\test\python\scratch.py", line 7, in g
    porcelain.pull('c:/Users/mattk/.test', 'https://github.com/3ll3d00d/test.git')
  File "C:\Users\mattk\Anaconda3_64\envs\beq\lib\site-packages\dulwich\porcelain.py", line 859, in pull
    path, r, progress=errstream.write, determine_wants=determine_wants)
  File "C:\Users\mattk\Anaconda3_64\envs\beq\lib\site-packages\dulwich\client.py", line 403, in fetch
    commit()
  File "C:\Users\mattk\Anaconda3_64\envs\beq\lib\site-packages\dulwich\client.py", line 389, in commit
    target.object_store.add_thin_pack(f.read, None)
  File "C:\Users\mattk\Anaconda3_64\envs\beq\lib\site-packages\dulwich\object_store.py", line 702, in add_thin_pack
    return self._complete_thin_pack(f, path, copier, indexer)
  File "C:\Users\mattk\Anaconda3_64\envs\beq\lib\site-packages\dulwich\object_store.py", line 672, in _complete_thin_pack
    index_file.close()
  File "C:\Users\mattk\Anaconda3_64\envs\beq\lib\site-packages\dulwich\file.py", line 174, in close
    os.replace(self._lockfilename, self._filename)
PermissionError: [WinError 5] Access is denied: 'c:/Users/mattk/.test\\.git\\objects\\pack\\pack-da39a3ee5e6b4b0d3255bfef95601890afd80709.idx.lock' -> 'c:/Users/mattk/.test\\.git\\objects\\pack\\pack-da39a3ee5e6b4b0d3255bfef95601890afd80709.idx'
```

NB: more often than not means sometimes it runs to completion happily

**Comment by bmiller** [comment]
I wonder if this could even be related to the #221 issue that I'm seeing.  On AVS you mentioned looking at other options.  Looking around all I'm seeing is either porcelain or pygit which you had dropped since it needs git installed.  Do you think making a git a dependency like ffmpeg might be worthwhile?  

For the mac build, it looks like I can get git and ffmpeg to build into the app by adding it to the binaries array in the .spec.  I haven't been able to get them to execute yet.  I think it's because they're not in the user's PATH when in the app bundle.  I'm looking into if it's possible and how to do it.  If this works then git could be bundled in too.  

The other thing I started playing with is a shell script that runs the two commands to get ffmpeg installed.  That would be easy to extend for git also.  Adding an "Install Tools" menu item that runs the script might be another option.  I know that'll work for Mac and Linux but I'm not sure about Windows.

**Comment by 3ll3d00d** [comment]
AFAIK the only pure python git impl is dulwich, the linked issue for #221 is one they say they are keen to fix for their 0.20 release so there is some hope that it will be sorted soon enough. The alternative is 
 going back to gitpython and telling people to install git for windows (or how to install git on osx/linux). At some point I need to start on some reference docs I guess....

bundling native binaries (or making an installer that grabs them) is opening the door to a support burden that I'm not interested in maintaining. If people can download beqd then they are capable of downloading ffmpeg or git :)

**Comment by bmiller** [comment]
True, support sucks.  I think I'll just get the install script working for mac.  All I was doing with it is running the command to install homebrew (package manager on mac) if it's not installed already and then calling brew to install ffmpeg.  I figured it would make things a little easier.  One of the guys using the mac build of BEQDesigner had no clue about using Terminal and said he thought he was in the matrix watching the logging of ffmpeg getting installed. :)
